commons-core/src/main/scala/com/avsystem/commons/serialization/json/JsonStringInput.scala

```diff
@@ -45,9 +43,9 @@ class JsonStringInput(reader: JsonReader, callback: AfterElement = AfterElementN
     case null => "null"
   }
 
-  private def expected(what: String) = throw new ReadFailure(s"Expected $what but got $jsonType")
+  private def expected(what: String) = throw new ReadFailure(s"Expected $what but got $jsonType: $value")
 
-  private def matchOr[T: ClassTag](what: String): T = value match {
+  private def matchOr[@specialized(Int, Double, Boolean) T: ClassTag](what: String): T = value match {
```

Comment:
Can you benchmark this?

Reply:
It looks like a small improvement on my machine:
```
commons-benchmark/jmh:run -i 10 -wi 4 -f1 -t1 .*JsonWriting.*
commons-benchmark/jmh:run -i 10 -wi 4 -f1 -t1 .*JsonReading.*
```

specialized
```
[info] JsonWritingBenchmark.writeCCGenCodec    thrpt   10  1770973.495 ± 32871.418  ops/s
[info] JsonWritingBenchmark.writeFoosGenCodec  thrpt   10     5408.593 ±   140.633  ops/s
[info] JsonWritingBenchmark.writeSHGenCodec    thrpt   10   902129.233 ± 43101.914  ops/s

[info] JsonReadingBenchmark.readCCGenCodec    thrpt   10  1104491.994 ± 62490.339  ops/s
[info] JsonReadingBenchmark.readFoosGenCodec  thrpt   10     3295.811 ±    46.822  ops/s
[info] JsonReadingBenchmark.readSHGenCodec    thrpt   10   570773.312 ±  7017.326  ops/s
```

non-specialized
```
[info] JsonWritingBenchmark.writeCCGenCodec    thrpt   10  1737726.156 ± 22873.874  ops/s
[info] JsonWritingBenchmark.writeFoosGenCodec  thrpt   10     5181.003 ±    62.474  ops/s
[info] JsonWritingBenchmark.writeSHGenCodec    thrpt   10   911119.979 ± 17322.355  ops/s

[info] JsonReadingBenchmark.readCCGenCodec    thrpt   10  1048688.625 ± 10050.140  ops/s
[info] JsonReadingBenchmark.readFoosGenCodec  thrpt   10     3255.139 ±    55.571  ops/s
[info] JsonReadingBenchmark.readSHGenCodec    thrpt   10   569297.300 ±  4471.684  ops/s
```